src/main/java/com/google/devtools/build/lib/vfs/AbstractFileSystem.java

```diff
@@ -48,35 +52,65 @@ protected InputStream getInputStream(Path path) throws IOException {
     // This loop is a workaround for an apparent bug in FileInputStream.open, which delegates
     // ultimately to JVM_Open in the Hotspot JVM.  This call is not EINTR-safe, so we must do the
     // retry here.
+    InputStream stream = null;
     for (; ; ) {
       try {
-        return createFileInputStream(path);
+        stream = createFileInputStream(path);
       } catch (FileNotFoundException e) {
         if (e.getMessage().endsWith("(Interrupted system call)")) {
           continue;
         } else {
           throw e;
         }
+      } catch (AccessDeniedException e) {
+        throw new IOException(e.getMessage() + " (Permission denied)", e);
       }
+      break;
     }
+
+    // On some platforms, whether a path can be read as a file won't be checked until
+    // the first read().
+    //
+    // http://mail.openjdk.java.net/pipermail/nio-dev/2014-December/002877.html
+    //
+    // The underlying stream might not be seekable, and wrapping in a BufferedInputStream
+    // causes test failures in the integration suite, so create our own little one-byte
+    // buffer here.
+    byte[] buf = new byte[1];
+    int buflen = 0;
+    try {
+      buflen = stream.read(buf, 0, 1);
```

Comment:
In what cases does this matter? From reading the above thread, it seems that this doesn't have anything to do with whether the path contains code points >128, does it?

Reply:
Nothing to do with encoding, this particular pile of nonsense is for the "no error until first read" issue from migrating `java.io` -> `java.nio`. There's a test that calling `getInputStream()` on a directory will fail.

This is reasonable! It should fail!

But in `java.nio` it won't fail, it will return a valid `InputStream` that throws an exception when the very first byte is read.

I have no idea whether this code is good -- the last time I wrote Java professionally was during the days of applets -- but it does allow the test to pass.

Reply:
Oh, I see. I missed the part that this is only a problem with `java.nio`.